src/verso/Verso/Doc/Concrete.lean

```diff
@@ -168,6 +171,61 @@ macro "%docName" moduleName:ident : term =>
 def currentDocName [Monad m] [MonadEnv m] : m Name := do
   pure <| docName <| (← Lean.MonadEnv.getEnv).mainModule
 
+open Language
+
+/--
+The complete state of `PartElabM`, suitable for saving and restoration during incremental
+elaboration
+-/
+structure DocElabSnapshotState where
+  docState : DocElabM.State
+  partState : PartElabM.State
+  termState : Term.SavedState
+deriving Nonempty
+
+structure DocElabSnapshotData extends Language.Snapshot where
+  stx      : Syntax
+  finished : Task DocElabSnapshotState
+deriving Nonempty
+
+inductive DocElabSnapshot where
+  | mk (data : DocElabSnapshotData) (next : Option (SnapshotTask DocElabSnapshot))
+deriving TypeName, Nonempty
```

Comment:
I think we could bake this part into the `incrementallyElabCommand` API in order to reduce the number of required accessor parameters without really losing generality. Or rather, we could if it wasn't for `TypeName`?

Reply:
I got about 70% of the way there.